std/socket.zig

```diff
@@ -0,0 +1,678 @@
+                    system.ERROR.WSAEADDRINUSE => return ConnectError.AddressInUse,
+                    system.ERROR.WSAEADDRNOTAVAIL => return ConnectError.AddressNotAvailable,
+                    system.ERROR.WSAEAFNOSUPPORT => return ConnectError.AddressFamilyNotSupported,
+                    system.ERROR.WSAEWOULDBLOCK => return ConnectError.system.emResources,
+                    system.ERROR.WSAEALREADY => unreachable, // The socket is nonblocking and a previous connection attempt has not yet been completed.
+                    system.ERROR.WSAEBADF => unreachable, // socket is not a valid open file descriptor.
+                    system.ERROR.WSAECONNREFUSED => return ConnectError.ConnectionRefused,
+                    system.ERROR.WSAEFAULT => unreachable, // The socket structure address is outside the user's address space.
+                    system.ERROR.WSAEISCONN => unreachable, // The socket is already connected.
+                    system.ERROR.WSAENETUNREACH => return ConnectError.NetworkUnreachable,
+                    system.ERROR.WSAENOTSOCK => unreachable, // The file descriptor socket does not refer to a socket.
+                    system.ERROR.WSAEPROTOTYPE => unreachable, // The socket type does not support the requested communications protocol.
+                    system.ERROR.WSAETIMEDOUT => return ConnectError.ConnectionTimedOut,
+                    else => return unexpectedError(@intCast(u32, err)),
+                },
+                else => return unexpectedError(@intCast(u32, err)),
+                system.ERROR.WSAEBADF => unreachable, // The argument socket is not a valid file descriptor.
+                system.ERROR.WSAEFAULT => unreachable, // The address pointed to by optval or optlen is not in a valid part of the process address space.
+                system.ERROR.WSAEINVAL => unreachable,
+                system.ERROR.WSAENOPROTOOPT => unreachable, // The option is unknown at the level indicated.
+                system.ERROR.WSAENOTSOCK => unreachable, // The file descriptor socket does not refer to a socket.
+            }
+        }
+    }
+
+    pub fn getSockName(self: Socket) GetSockNameError!net.Address {
+        var addr: net.OsAddress = undefined;
+        var addrlen: system.socklen_t = @sizeOf(net.OsAddress);
+        const rc = system.getsockname(self.fd, &addr, &addrlen);
+
+        if (is_posix) {
+            const err = system.getErrno(rc);
+            switch (err) {
+                0 => return net.Address.init(addr),
+                else => return unexpectedError(err),
+                system.EBADF => unreachable,
+                system.EFAULT => unreachable,
+                system.EINVAL => unreachable,
+                system.ENOTSOCK => return GetSockNameError.FileDescriptorNotASocket,
+                system.ENOBUFS => return GetSockNameError.system.emResources,
+            }
+        } else {
+            const err = system.WSAGetLastError();
```

Comment:
You should only call WSAGetLastERror if the windows function you just called indicated there was an error.  Otherwise, WSAGetLastError could be returning a non-zero value from a previous function that failed or a previous function that sets the "last error" to an non-zero value for extra information to the caller.